### PR TITLE
DQN v2 Rewards & Algorithm Changes

### LINKED ISSUE #59: Update Rewards for DQN V2
We need to be a better tutor, update rewards methodology to be better stimulate learning within the agents.

### PR DESCRIPTION
**Pull request by Thanoshan**
This PR merges in the changes that enable our DQN v2 Flagship Algorithm. We also have changes to the reward functions including, individualized feedback, delta rewards and time penalty rewards.

This PR closes issue #59

### CODE REVIEW

**Review by pandyah5 — CHANGES_REQUESTED** [review]
LGTM overall, I have left a few comments on specific sections

File: src/nepiada/rl_ma_dqn_experimental.py [hunk 1]
```diff
@@ -24,7 +24,9 @@ def on_sample_end(self, *, worker: RolloutWorker, samples: SampleBatch, **kwargs
             curr_epsilon = worker.policy_map[curr_agent_name].get_exploration_state()["cur_epsilon"]
             curr_timestep = worker.policy_map[curr_agent_name].get_exploration_state()["last_timestep"]
             agent_epsilons += f"{curr_agent_name}: {str(curr_epsilon)} | {str(curr_timestep)} || "
-
+        # curr_epsilon = worker.policy_map["parameter_sharing"].get_exploration_state()["cur_epsilon"]
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/rl_ma_dqn_experimental.py:27]: Yeah you would need this if you do parameter sharing, I'd keep this in.

File: src/nepiada/rl_ma_dqn_experimental.py [hunk 2]
```diff
@@ -52,18 +54,17 @@ def env_creator(args):
 
     config = config.training(replay_buffer_config=replay_config, num_atoms=1, gamma=0.5, lr=0.0005)
     config = config.resources(num_gpus=1)
-    config = config.rollouts(num_rollout_workers=15, rollout_fragment_length=100, compress_observations=True)
+    config = config.rollouts(num_rollout_workers=0, rollout_fragment_length=100, compress_observations=True)
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/rl_ma_dqn_experimental.py:57]: Yeah you need multiple workers for sure for training, I like to use no rollout workers during demo/testing thourhg post training cause you don't need samplers. I'd leave it at 0 and we can change it to whatever on our own end depending on our system specs.

File: src/nepiada/env/nepiada.py [hunk 3]
```diff
@@ -574,7 +575,16 @@ def get_rewards(self):
         curr_scores = self._compute_scores()
 
         for agent_name in self.agents:
-            rewards[agent_name] = curr_scores[agent_name]
+            rewards[agent_name] = curr_scores[agent_name] - self.world.agents[agent_name].prev_score
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/env/nepiada.py:578]: Done.

File: src/nepiada/env/nepiada.py [hunk 4]
```diff
@@ -574,7 +575,16 @@ def get_rewards(self):
...
+            # Add timestep penalty, make it more negative if rewards is calculated to be negative, and less positive if rewards calculated to be positive
+            if rewards[agent_name] < 0:
+                # Up to -2x the reward based on num_moves and iterations
+                rewards[agent_name] = rewards[agent_name] + ((self.num_moves / self.config.iterations) * rewards[agent_name])
+            elif rewards[agent_name] > 0:
+                # Up to -2x the reward based on num_moves and iterations
+                rewards[agent_name] = rewards[agent_name] - ((self.num_moves / self.config.iterations) * rewards[agent_name])
+                if rewards[agent_name] < 0:
```
[the remark this answers is not in the document]
Comment (Thanoshan) [reply src/nepiada/env/nepiada.py]: You're right, this would not be triggered ever. IT was a relic of an older function I was trying. Removed it.

**Review by pandyah5 — APPROVED** [review]

**Pull request merged**